### PR TITLE
(proof-new) Improve interfaces to proof generators

### PR DESCRIPTION
**Pull request by ajreynol**
This includes configurable naming and a caching policy for term conversion proof generator.

Also corrects a subtle issue in LazyCDProof related to making getProofFor idempotent using the notion of owned proofs.

### CODE REVIEW

File: src/expr/proof.h [hunk 1]
```diff
@@ -241,8 +245,8 @@ class CDProof : public ProofGenerator
   context::Context d_context;
   /** The nodes of the proof */
   NodeProofNodeMap d_nodes;
-  /** Get proof for fact, or nullptr if it does not exist. */
-  std::shared_ptr<ProofNode> getProof(Node fact) const;
+  /** Name dentifier */
```
Comment (HanielB) [inline src/expr/proof.h:248]:
```suggestion
  /** Name identifier */
```

File: src/theory/eager_proof_generator.h [hunk 2]
```diff
@@ -163,6 +165,8 @@ class EagerProofGenerator : public ProofGenerator
   void setProofForPropExp(TNode lit, Node exp, std::shared_ptr<ProofNode> pf);
   /** The proof node manager */
   ProofNodeManager* d_pnm;
+  /** Name dentifier */
```
Comment (HanielB) [inline src/theory/eager_proof_generator.h:168]:
```suggestion
  /** Name identifier */
```

File: src/smt/term_formula_removal.h [hunk 3]
```diff
@@ -106,10 +106,10 @@ class RemoveTermFormulas {
   void garbageCollect();
 
   /**
-   * Set proof checker, which signals this class to enable proofs using the
+   * Set proof node manager, which signals this class to enable proofs using the
    * given checker.
```
Comment (HanielB) [inline src/smt/term_formula_removal.h:110]: update

File: src/smt/term_formula_removal.h [hunk 4]
```diff
@@ -166,7 +166,7 @@ class RemoveTermFormulas {
   static bool hasNestedTermChildren( TNode node );
 
   /** A proof node manager */
-  std::unique_ptr<ProofNodeManager> d_pnm;
+  ProofNodeManager* d_pnm;
```
Comment (HanielB) [inline src/smt/term_formula_removal.h:169]: why did you change this?

File: src/expr/term_conversion_proof_generator.h [hunk 5]
```diff
@@ -35,6 +35,19 @@ enum class TConvPolicy : uint32_t
 /** Writes a term conversion policy name to a stream. */
 std::ostream& operator<<(std::ostream& out, TConvPolicy tcpol);
 
+/** A policy for how proofs are cache in TConvProofGenerator */
```
Comment (HanielB) [inline src/expr/term_conversion_proof_generator.h:38]:
```suggestion
/** A policy for how proofs are cached in TConvProofGenerator */
```

File: src/expr/term_conversion_proof_generator.h [hunk 6]
```diff
@@ -85,7 +98,9 @@ class TConvProofGenerator : public ProofGenerator
    */
   TConvProofGenerator(ProofNodeManager* pnm,
                       context::Context* c = nullptr,
-                      TConvPolicy pol = TConvPolicy::FIXPOINT);
+                      TConvPolicy pol = TConvPolicy::FIXPOINT,
+                      TConvCachePolicy cpol = TConvCachePolicy::NEVER,
```
Comment (HanielB) [inline src/expr/term_conversion_proof_generator.h:102]: in the documentation above add `cpol`. Also say why the defaults for `pol` and `cpol` are these.

File: src/expr/proof_generator.h [hunk 7]
```diff
@@ -105,6 +105,30 @@ class ProofGenerator
   virtual std::string identify() const = 0;
 };
 
+/**
+ * debug check closed on Trace c, context ctx is string for debugging
```
Comment (HanielB) [inline src/expr/proof_generator.h:109]:
```suggestion
 * Debug check closed with Trace c. Context ctx is string for debugging.
```

File: src/expr/proof_generator.h [hunk 8]
```diff
@@ -105,6 +105,30 @@ class ProofGenerator
...
+ * assumps is the set of allowed open assertions.
+ *
+ * @param reqGen Whether we consider a null generator to be a failure.
+ */
+void pfgEnsureClosedWrt(Node proven,
+                     ProofGenerator* pg,
+                     const std::vector<Node>& assumps,
+                     const char* c,
+                     const char* ctx,
+                     bool reqGen = true);
```
Comment (HanielB) [inline src/expr/proof_generator.h:130]: format

File: src/expr/proof_generator.h [hunk 9]
```diff
@@ -105,6 +105,30 @@ class ProofGenerator
...
+ */
+void pfgEnsureClosed(Node proven,
+                     ProofGenerator* pg,
+                     const char* c,
+                     const char* ctx,
+                     bool reqGen = true);
+
+/**
+ * debug check closed on Trace c, context ctx is string for debugging, where
+ * assumps is the set of allowed open assertions.
```
Comment (HanielB) [inline src/expr/proof_generator.h:121]:
```suggestion
 * Debug check closed with Trace c. Context ctx is string for debugging and
 * assumps is the set of allowed open assertions.
```

File: src/expr/proof_generator.cpp [hunk 10]
```diff
@@ -66,4 +68,94 @@ bool ProofGenerator::addProofTo(Node f, CDProof* pf, CDPOverwrite opolicy)
...
+  {
+    // proofs not enabled, do not do check
+    return;
+  }
+  bool isTraceDebug = Trace.isOn(c);
+  if (!options::proofNewEagerChecking() && !isTraceDebug)
+  {
+    // trace is off and proof new eager checking is off, do not do check
+    return;
+  }
```
Comment (HanielB) [inline src/expr/proof_generator.cpp:98]: Do we need these hard coded tests here? It'd make more sense to me to test this in the caller.

File: src/expr/proof_generator.cpp [hunk 11]
```diff
@@ -66,4 +68,94 @@ bool ProofGenerator::addProofTo(Node f, CDProof* pf, CDPOverwrite opolicy)
...
+    }
+    Trace(c) << "...pfgEnsureClosed: no generator in context " << ctx
+             << std::endl;
+    return;
+  }
+  std::shared_ptr<ProofNode> pn = pg->getProofFor(proven);
+  Trace(c) << " Proof: " << *pn.get() << std::endl;
+  if (pn == nullptr)
+  {
+    AlwaysAssert(false) << "...pfgEnsureClosed: null proof from " << ss.str()
```
Comment (HanielB) [inline src/expr/proof_generator.cpp:126]: use `Unreachable`

File: src/expr/proof_generator.cpp [hunk 12]
```diff
@@ -66,4 +68,94 @@ bool ProofGenerator::addProofTo(Node f, CDProof* pf, CDPOverwrite opolicy)
...
+    sdiag << ", use -t " << c << " for details";
+  }
+  Trace(c) << "=== pfgEnsureClosed: " << ss.str() << std::endl;
+  Trace(c) << "Proven: " << proven << std::endl;
+  if (pg == nullptr)
+  {
+    // only failure if flag is true
+    if (reqGen)
+    {
+      AlwaysAssert(false) << "...pfgEnsureClosed: no generator in context "
```
Comment (HanielB) [inline src/expr/proof_generator.cpp:115]: use `Unreachable`

File: src/expr/proof_generator.cpp [hunk 13]
```diff
@@ -66,4 +68,94 @@ bool ProofGenerator::addProofTo(Node f, CDProof* pf, CDPOverwrite opolicy)
...
+    AlwaysAssert(false) << "...pfgEnsureClosed: null proof from " << ss.str()
+                        << sdiag.str();
+  }
+  std::vector<Node> fassumps;
+  expr::getFreeAssumptions(pn.get(), fassumps);
+  bool isClosed = true;
+  std::stringstream ssf;
+  for (const Node& fa : fassumps)
+  {
+    if (std::find(assumps.begin(),assumps.end(),fa)==assumps.end())
```
Comment (HanielB) [inline src/expr/proof_generator.cpp:135]: we should be checking modulo symmetry, no?

File: src/expr/proof_generator.h [hunk 14]
```diff
@@ -105,6 +105,30 @@ class ProofGenerator
   virtual std::string identify() const = 0;
 };
 
+/**
+ * debug check closed on Trace c, context ctx is string for debugging
+ *
+ * @param reqGen Whether we consider a null generator to be a failure.
+ */
+void pfgEnsureClosed(Node proven,
```
Comment (HanielB) [inline src/expr/proof_generator.h:113]: Please add to the doc above the policy for whether this method being a no-op. It's unclear as is.

File: src/expr/lazy_proof.cpp [hunk 15]
```diff
@@ -56,36 +57,41 @@ std::shared_ptr<ProofNode> LazyCDProof::getProofFor(Node fact)
     if (it == visited.end())
     {
       visited.insert(cur);
-      if (cur->getRule() == PfRule::ASSUME)
+      Node cfact = cur->getResult();
+      if (getProof(cfact).get() != cur)
+      {
+        // we don't own this proof, skip it
```
Comment (HanielB) [inline src/expr/lazy_proof.cpp:63]: Clarify saying that this is so that the method is idempotent and does not expand assumptions that were added from previous expansions of generators.

File: src/expr/lazy_proof.cpp [hunk 16]
```diff
@@ -56,36 +57,41 @@ std::shared_ptr<ProofNode> LazyCDProof::getProofFor(Node fact)
...
-          Assert(!afactGen.isNull());
-          // use the addProofTo interface
-          if (!pg->addProofTo(afactGen, this))
+          Trace("lazy-cdproof")
+              << "LazyCDProof: Call generator " << pg->identify()
+              << " for assumption " << cfact << std::endl;
+          Node cfactGen = isSym ? CDProof::getSymmFact(cfact) : cfact;
+          Assert(!cfactGen.isNull());
+          // do not use the addProofTo interface
+          // instead use the update node interface
```
Comment (HanielB) [inline src/expr/lazy_proof.cpp:78]: say why

File: src/expr/lazy_proof.cpp [hunk 17]
```diff
@@ -104,14 +110,34 @@ std::shared_ptr<ProofNode> LazyCDProof::getProofFor(Node fact)
...
+                              bool forceOverwrite,
+                              PfRule idNull)
 {
-  Assert(pg != nullptr);
+  if (pg == nullptr)
+  {
+    // null generator, should have given a proof rule
+    if (idNull == PfRule::ASSUME)
+    {
+      AlwaysAssert(false) << "LazyCDProof::addLazyStep: " << identify()
```
Comment (HanielB) [inline src/expr/lazy_proof.cpp:129]: Use `Unreachable`

File: src/expr/proof_generator.cpp (hunk 10 above)
Comment (ajreynol) [reply src/expr/proof_generator.cpp:98]: Yes we need them, or else this is a major pain to do over and over in the code, which is what I had before.  The caller should not be responsible for getting this behavior right.

File: src/expr/proof_generator.cpp (hunk 13 above)
Comment (ajreynol) [reply src/expr/proof_generator.cpp:135]: No, that is specific to CDProof.

File: src/smt/term_formula_removal.h (hunk 4 above)
Comment (ajreynol) [reply src/smt/term_formula_removal.h:169]: Because we will use a single global proof node manager, objects should not create their own.

File: src/expr/term_conversion_proof_generator.h (hunk 6 above)
Comment (HanielB) [reply src/expr/term_conversion_proof_generator.h:104]: You did not explain the defaults.

Comment (ajreynol) [reply src/expr/term_conversion_proof_generator.h:106]: Done
